op-node/rollup/derive/payload_util.go

```diff
@@ -59,34 +60,41 @@ func PayloadToSystemConfig(rollupCfg *rollup.Config, payload *eth.ExecutionPaylo
 				rollupCfg.Genesis.L2.Number, payload.BlockHash, rollupCfg.Genesis.L2.Hash)
 		}
 		return rollupCfg.Genesis.SystemConfig, nil
-	} else {
-		if len(payload.Transactions) == 0 {
-			return eth.SystemConfig{}, fmt.Errorf("l2 block is missing L1 info deposit tx, block hash: %s", payload.BlockHash)
-		}
-		var tx types.Transaction
-		if err := tx.UnmarshalBinary(payload.Transactions[0]); err != nil {
-			return eth.SystemConfig{}, fmt.Errorf("failed to decode first tx to read l1 info from: %w", err)
-		}
-		if tx.Type() != types.DepositTxType {
-			return eth.SystemConfig{}, fmt.Errorf("first payload tx has unexpected tx type: %d", tx.Type())
-		}
-		info, err := L1BlockInfoFromBytes(rollupCfg, uint64(payload.Timestamp), tx.Data())
-		if err != nil {
-			return eth.SystemConfig{}, fmt.Errorf("failed to parse L1 info deposit tx from L2 block: %w", err)
-		}
-		if isEcotoneButNotFirstBlock(rollupCfg, uint64(payload.Timestamp)) {
-			// Translate Ecotone values back into encoded scalar if needed.
-			// We do not know if it was derived from a v0 or v1 scalar,
-			// but v1 is fine, a 0 blob base fee has the same effect.
-			info.L1FeeScalar[0] = 1
-			binary.BigEndian.PutUint32(info.L1FeeScalar[24:28], info.BlobBaseFeeScalar)
-			binary.BigEndian.PutUint32(info.L1FeeScalar[28:32], info.BaseFeeScalar)
+	}
+
+	if len(payload.Transactions) == 0 {
+		return eth.SystemConfig{}, fmt.Errorf("l2 block is missing L1 info deposit tx, block hash: %s", payload.BlockHash)
+	}
+	var tx types.Transaction
+	if err := tx.UnmarshalBinary(payload.Transactions[0]); err != nil {
+		return eth.SystemConfig{}, fmt.Errorf("failed to decode first tx to read l1 info from: %w", err)
+	}
+	if tx.Type() != types.DepositTxType {
+		return eth.SystemConfig{}, fmt.Errorf("first payload tx has unexpected tx type: %d", tx.Type())
+	}
+	info, err := L1BlockInfoFromBytes(rollupCfg, uint64(payload.Timestamp), tx.Data())
+	if err != nil {
+		return eth.SystemConfig{}, fmt.Errorf("failed to parse L1 info deposit tx from L2 block: %w", err)
+	}
+	if isEcotoneButNotFirstBlock(rollupCfg, uint64(payload.Timestamp)) {
+		// Translate Ecotone values back into encoded scalar if needed.
+		// We do not know if it was derived from a v0 or v1 scalar,
+		// but v1 is fine, a 0 blob base fee has the same effect.
+		info.L1FeeScalar[0] = 1
+		binary.BigEndian.PutUint32(info.L1FeeScalar[24:28], info.BlobBaseFeeScalar)
+		binary.BigEndian.PutUint32(info.L1FeeScalar[28:32], info.BaseFeeScalar)
+	}
+	r := eth.SystemConfig{
+		BatcherAddr: info.BatcherAddr,
+		Overhead:    info.L1FeeOverhead,
+		Scalar:      info.L1FeeScalar,
+		GasLimit:    uint64(payload.GasLimit),
+	}
+	if rollupCfg.IsHolocene(uint64(payload.Timestamp)) {
+		if err := eip1559.ValidateHoloceneExtraData(payload.ExtraData); err != nil {
```

Comment:
We could change this into a `ParseHoloceneExtraData([]byte) ([8]byte, error)` or so, to encapsulate the parsing logic and remove the footgun to throw away the version prefix.

Reply:
Originally I had done the following to avoid this abstraction violation, but then thought it was unnecessarily verbose.  Could definitely add a method like you suggest but it would best be put in op-geth I think?

```
if err := eip1559.validate...
d, e := eip1559.DecodeHoloceneExtraData(...
s := eip1559.EncodeHolocene1559Params(d, e)
copy(r.EIP1559...
```

Reply:
changed it back to the above to avoid the footgun.

Reply:
nice that works too